File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio 9.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <limits>
#include <ctime>

const std::string ARCHIVO_VENTAS = "ventas.txt";
const std::string ARCHIVO_FACTURAS = "facturas.txt";

class Venta {
public:
    Venta(int numeroFactura, const std::string &producto, int cantidad, double precio)
        : numeroFactura(numeroFactura), producto(producto), cantidad(cantidad), precio(precio) {
        fecha = obtenerFechaActual();
    }

    int getNumeroFactura() const { return numeroFactura; }
    const std::string &getProducto() const { return producto; }
    int getCantidad() const { return cantidad; }
    double getPrecio() const { return precio; }
    const std::string &getFecha() const { return fecha; }

    void setFecha(const std::string &fechaValue) {
        fecha = fechaValue;
    }

    double calcularTotal() const {
        return cantidad * precio;
    }

    std::string generarFactura() const {
        std::ostringstream out;
        out << "+------------------------------------------------+\n";
        out << "¦            FACTURA DE VENTA\n";
        out << "+------------------------------------------------+\n";
        out << "Número de Factura: " << numeroFactura << "\n";
        out << "Fecha: " << fecha << "\n";
        out << "------------------------------------------------\n";
        out << "Producto: " << producto << "\n";
        out << "Cantidad: " << cantidad << "\n";
        out << "Precio Unitario: $" << precio << "\n";
        out << "------------------------------------------------\n";
        out << "TOTAL: $" << calcularTotal() << "\n";
        out << "------------------------------------------------\n";
        return out.str();
    }

    std::string toRegistro() const {
        std::ostringstream out;
        out << numeroFactura << "|" << producto << "|" << cantidad << "|" << precio << "|" << fecha;
        return out.str();
    }

private:
    int numeroFactura;
    std::string producto;
    int cantidad;
    double precio;
    std::string fecha;

    std::string obtenerFechaActual() const {
        std::time_t now = std::time(nullptr);
        char buffer[32];
        std::tm localTime;
#if defined(_WIN32) || defined(_WIN64)
        localtime_s(&localTime, &now);
#else
        localtime_r(&now, &localTime);
#endif
        std::strftime(buffer, sizeof(buffer), "%d/%m/%Y %H:%M:%S", &localTime);
        return std::string(buffer);
    }
};

int leerEntero(const std::string &mensaje) {
    while (true) {
        std::cout << mensaje;
        int valor;
        if (std::cin >> valor) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return valor;
        }
        std::cout << "Debe ingresar un número válido." << std::endl;
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
}

double leerDoublePositivo(const std::string &mensaje) {
    while (true) {
        std::cout << mensaje;
        double valor;
        if (std::cin >> valor) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            if (valor > 0) {
                return valor;
            }
            std::cout << "Debe ingresar un valor mayor que cero." << std::endl;
        } else {
            std::cout << "Debe ingresar un número válido." << std::endl;
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        }
    }
}

std::string leerTextoNoVacio(const std::string &mensaje) {
    while (true) {
        std::cout << mensaje;
        std::string texto;
        std::getline(std::cin, texto);
        if (!texto.empty()) {
            return texto;
        }
        std::cout << "El texto no puede estar vacío." << std::endl;
    }
}

void guardarFactura(const std::string &factura) {
    std::ofstream archivo(ARCHIVO_FACTURAS, std::ios::app);
    if (!archivo) {
        std::cout << "Error al guardar factura." << std::endl;
        return;
    }
    archivo << factura << "\n" << std::string(50, '-') << "\n";
}

std::vector<Venta> cargarVentas(int &numeroFactura) {
    std::vector<Venta> ventas;
    std::ifstream archivo(ARCHIVO_VENTAS);
    if (!archivo) {
        return ventas;
    }

    std::string linea;
    while (std::getline(archivo, linea)) {
        std::istringstream ss(linea);
        std::string numeroStr, producto, cantidadStr, precioStr, fecha;

        if (std::getline(ss, numeroStr, '|') && std::getline(ss, producto, '|') &&
            std::getline(ss, cantidadStr, '|') && std::getline(ss, precioStr, '|') &&
            std::getline(ss, fecha)) {

            try {
                int numero = std::stoi(numeroStr);
                int cantidad = std::stoi(cantidadStr);
                double precio = std::stod(precioStr);
                Venta venta(numero, producto, cantidad, precio);
                venta.setFecha(fecha);
                ventas.push_back(venta);
                if (numero >= numeroFactura) {
                    numeroFactura = numero + 1;
                }
            } catch (...) {
                // ignorar línea malformada
            }
        }
    }

    return ventas;
}

void guardarVentas(const std::vector<Venta> &ventas) {
    std::ofstream archivo(ARCHIVO_VENTAS);
    if (!archivo) {
        std::cout << "Error al guardar ventas." << std::endl;
        return;
    }
    for (const auto &venta : ventas) {
        archivo << venta.toRegistro() << "\n";
    }
}

Venta *buscarVentaPorNumero(std::vector<Venta> &ventas, int numero) {
    for (auto &venta : ventas) {
        if (venta.getNumeroFactura() == numero) {
            return &venta;
        }
    }
    return nullptr;
}

std::string truncar(const std::string &texto, size_t longitud) {
    if (texto.size() <= longitud) {
        return texto;
    }
    return texto.substr(0, longitud - 2) + "..";
}

void mostrarMenu() {
    std::cout << "=== Sistema de Ventas ===" << std::endl;
    std::cout << "1. Registrar venta" << std::endl;
    std::cout << "2. Generar factura" << std::endl;
    std::cout << "3. Mostrar historial de ventas" << std::endl;
    std::cout << "4. Ver estadísticas" << std::endl;
    std::cout << "5. Salir" << std::endl;
}

void mostrarHistorialVentas(const std::vector<Venta> &ventas) {
    if (ventas.empty()) {
        std::cout << "No hay ventas registradas." << std::endl;
        return;
    }

    std::cout << "=== Historial de Ventas ===" << std::endl;
    std::cout << std::left << std::setw(5) << "ID" << std::setw(25) << "Producto"
              << std::setw(10) << "Cantidad" << std::setw(14) << "Precio Unit." << std::setw(14)
              << "Total" << std::endl;
    std::cout << std::string(70, '-') << std::endl;

    for (const auto &venta : ventas) {
        std::cout << std::left << std::setw(5) << venta.getNumeroFactura()
                  << std::setw(25) << truncar(venta.getProducto(), 23)
                  << std::setw(10) << venta.getCantidad()
                  << "$" << std::setw(13) << std::fixed << std::setprecision(2) << venta.getPrecio()
                  << "$" << std::setw(13) << std::fixed << std::setprecision(2) << venta.calcularTotal()
                  << std::endl;
    }
}

void mostrarEstadisticas(const std::vector<Venta> &ventas) {
    if (ventas.empty()) {
        std::cout << "No hay ventas registradas." << std::endl;
        return;
    }

    double totalVentas = 0.0;
    int totalProductos = 0;
    for (const auto &venta : ventas) {
        totalVentas += venta.calcularTotal();
        totalProductos += venta.getCantidad();
    }
    double promedio = totalVentas / ventas.size();

    std::cout << "=== Estadísticas de Ventas ===" << std::endl;
    std::cout << "Total de transacciones: " << ventas.size() << std::endl;
    std::cout << "Total de productos vendidos: " << totalProductos << std::endl;
    std::cout << "Monto total vendido: $" << std::fixed << std::setprecision(2) << totalVentas << std::endl;
    std::cout << "Venta promedio: $" << std::fixed << std::setprecision(2) << promedio << std::endl;
}

int main() {
    int numeroFactura = 1;
    std::vector<Venta> ventas = cargarVentas(numeroFactura);

    int opcion;
    do {
        mostrarMenu();
        opcion = leerEntero("Seleccione una opción: ");
        switch (opcion) {
            case 1: {
                std::string producto = leerTextoNoVacio("Nombre del producto: ");
                int cantidad = leerEntero("Cantidad: ");
                double precio = leerDoublePositivo("Precio unitario: $");
                Venta venta(numeroFactura, producto, cantidad, precio);
                ventas.push_back(venta);
                numeroFactura++;
                guardarVentas(ventas);
                std::cout << "Venta registrada. Total: $" << std::fixed << std::setprecision(2)
                          << venta.calcularTotal() << std::endl;
                break;
            }
            case 2: {
                if (ventas.empty()) {
                    std::cout << "No hay ventas registradas." << std::endl;
                    break;
                }
                int numero = leerEntero("Número de factura: ");
                Venta *venta = buscarVentaPorNumero(ventas, numero);
                if (venta == nullptr) {
                    std::cout << "Factura no encontrada." << std::endl;
                } else {
                    std::string factura = venta->generarFactura();
                    guardarFactura(factura);
                    std::cout << factura << std::endl;
                    std::cout << "Factura guardada en " << ARCHIVO_FACTURAS << std::endl;
                }
                break;
            }
            case 3:
                mostrarHistorialVentas(ventas);
                break;
            case 4:
                mostrarEstadisticas(ventas);
                break;
            case 5:
                std::cout << "Gracias por usar el Sistema de Ventas. ¡Hasta luego!" << std::endl;
                break;
            default:
                std::cout << "Opción inválida. Intente de nuevo." << std::endl;
        }
        std::cout << std::endl;
    } while (opcion != 5);

    guardarVentas(ventas);
    return 0;
}

File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio9.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Ejercicio9 {

    private static final String ARCHIVO_VENTAS = "ventas.txt";
    private static final String ARCHIVO_FACTURAS = "facturas.txt";
    private static List<Venta> ventas = new ArrayList<>();
    private static int numeroFactura = 1;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        cargarVentas();

        int opcion;
        do {
            mostrarMenu();
            opcion = leerEntero(scanner, "Seleccione una opción: ");
            switch (opcion) {
                case 1:
                    registrarVenta(scanner);
                    break;
                case 2:
                    generarFactura(scanner);
                    break;
                case 3:
                    mostrarHistorialVentas();
                    break;
                case 4:
                    mostrarEstadisticas();
                    break;
                case 5:
                    System.out.println("Gracias por usar el Sistema de Ventas. ¡Hasta luego!");
                    break;
                default:
                    System.out.println("Opción inválida. Intente de nuevo.");
            }
            System.out.println();
        } while (opcion != 5);

        guardarVentas();
        scanner.close();
    }

    private static void mostrarMenu() {
        System.out.println("=== Sistema de Ventas ===");
        System.out.println("1. Registrar venta");
        System.out.println("2. Generar factura");
        System.out.println("3. Mostrar historial de ventas");
        System.out.println("4. Ver estadísticas");
        System.out.println("5. Salir");
    }

    private static int leerEntero(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String linea = scanner.nextLine().trim();
            try {
                return Integer.parseInt(linea);
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un número válido.");
            }
        }
    }

    private static double leerDoublePositivo(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String linea = scanner.nextLine().trim();
            try {
                double valor = Double.parseDouble(linea);
                if (valor <= 0) {
                    System.out.println("Debe ingresar un valor mayor que cero.");
                } else {
                    return valor;
                }
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un número válido.");
            }
        }
    }

    private static String leerTextoNoVacio(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String texto = scanner.nextLine().trim();
            if (!texto.isEmpty()) {
                return texto;
            }
            System.out.println("El texto no puede estar vacío.");
        }
    }

    private static void registrarVenta(Scanner scanner) {
        String producto = leerTextoNoVacio(scanner, "Nombre del producto: ");
        int cantidad = leerEntero(scanner, "Cantidad: ");
        double precio = leerDoublePositivo(scanner, "Precio unitario: $");

        Venta venta = new Venta(numeroFactura, producto, cantidad, precio);
        ventas.add(venta);
        numeroFactura++;

        guardarVentas();
        System.out.printf("Venta registrada. Total: $%.2f%n", venta.calcularTotal());
    }

    private static void generarFactura(Scanner scanner) {
        if (ventas.isEmpty()) {
            System.out.println("No hay ventas registradas.");
            return;
        }

        System.out.println("Seleccione el número de factura a generar:");
        int numero = leerEntero(scanner, "Número de factura: ");

        Venta venta = buscarVentaPorNumero(numero);
        if (venta != null) {
            String factura = venta.generarFactura();
            guardarFactura(factura);
            System.out.println(factura);
            System.out.println("Factura guardada en " + ARCHIVO_FACTURAS);
        } else {
            System.out.println("Factura no encontrada.");
        }
    }

    private static Venta buscarVentaPorNumero(int numero) {
        for (Venta venta : ventas) {
            if (venta.getNumeroFactura() == numero) {
                return venta;
            }
        }
        return null;
    }

    private static void mostrarHistorialVentas() {
        if (ventas.isEmpty()) {
            System.out.println("No hay ventas registradas.");
            return;
        }

        System.out.println("=== Historial de Ventas ===");
        System.out.printf("%-5s %-25s %-10s %-12s %-12s%n", "ID", "Producto", "Cantidad", "Precio Unit.", "Total");
        System.out.println("─".repeat(70));

        for (Venta venta : ventas) {
            System.out.printf("%-5d %-25s %-10d $%-11.2f $%-11.2f%n",
                    venta.getNumeroFactura(),
                    truncar(venta.getProducto(), 23),
                    venta.getCantidad(),
                    venta.getPrecio(),
                    venta.calcularTotal());
        }
    }

    private static void mostrarEstadisticas() {
        if (ventas.isEmpty()) {
            System.out.println("No hay ventas registradas.");
            return;
        }

        double totalVentas = 0;
        int totalProductos = 0;

        for (Venta venta : ventas) {
            totalVentas += venta.calcularTotal();
            totalProductos += venta.getCantidad();
        }

        double promedio = totalVentas / ventas.size();

        System.out.println("=== Estadísticas de Ventas ===");
        System.out.printf("Total de transacciones: %d%n", ventas.size());
        System.out.printf("Total de productos vendidos: %d%n", totalProductos);
        System.out.printf("Monto total vendido: $%.2f%n", totalVentas);
        System.out.printf("Venta promedio: $%.2f%n", promedio);
    }

    private static String truncar(String texto, int longitud) {
        if (texto.length() <= longitud) {
            return texto;
        }
        return texto.substring(0, longitud - 2) + "..";
    }

    private static void guardarFactura(String factura) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(ARCHIVO_FACTURAS, true))) {
            writer.write(factura);
            writer.write("\n" + "═".repeat(50) + "\n");
        } catch (IOException e) {
            System.out.println("Error al guardar factura: " + e.getMessage());
        }
    }

    private static void cargarVentas() {
        File archivo = new File(ARCHIVO_VENTAS);
        if (!archivo.exists()) {
            return;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(archivo))) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                String[] partes = linea.split("\\|");
                if (partes.length == 5) {
                    int numero = Integer.parseInt(partes[0]);
                    String producto = partes[1];
                    int cantidad = Integer.parseInt(partes[2]);
                    double precio = Double.parseDouble(partes[3]);
                    String fecha = partes[4];

                    Venta venta = new Venta(numero, producto, cantidad, precio);
                    venta.setFecha(fecha);
                    ventas.add(venta);

                    if (numero >= numeroFactura) {
                        numeroFactura = numero + 1;
                    }
                }
            }
        } catch (IOException | NumberFormatException e) {
            System.out.println("Error al cargar ventas: " + e.getMessage());
        }
    }

    private static void guardarVentas() {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(ARCHIVO_VENTAS))) {
            for (Venta venta : ventas) {
                writer.write(venta.toRegistro());
                writer.newLine();
            }
        } catch (IOException e) {
            System.out.println("Error al guardar ventas: " + e.getMessage());
        }
    }
}

class Venta {

    private final int numeroFactura;
    private final String producto;
    private final int cantidad;
    private final double precio;
    private String fecha;

    public Venta(int numeroFactura, String producto, int cantidad, double precio) {
        this.numeroFactura = numeroFactura;
        this.producto = producto;
        this.cantidad = cantidad;
        this.precio = precio;
        this.fecha = obtenerFechaActual();
    }

    public int getNumeroFactura() {
        return numeroFactura;
    }

    public String getProducto() {
        return producto;
    }

    public int getCantidad() {
        return cantidad;
    }

    public double getPrecio() {
        return precio;
    }

    public double calcularTotal() {
        return cantidad * precio;
    }

    public void setFecha(String fecha) {
        this.fecha = fecha;
    }

    private String obtenerFechaActual() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss");
        return LocalDateTime.now().format(formatter);
    }

    public String generarFactura() {
        StringBuilder factura = new StringBuilder();
        factura.append("╔════════════════════════════════════════════════╗\n");
        factura.append("║            FACTURA DE VENTA\n");
        factura.append("╚════════════════════════════════════════════════╝\n");
        factura.append(String.format("Número de Factura: %d\n", numeroFactura));
        factura.append(String.format("Fecha: %s\n", fecha));
        factura.append("────────────────────────────────────────────────\n");
        factura.append(String.format("Producto: %s\n", producto));
        factura.append(String.format("Cantidad: %d\n", cantidad));
        factura.append(String.format("Precio Unitario: $%.2f\n", precio));
        factura.append("────────────────────────────────────────────────\n");
        factura.append(String.format("TOTAL: $%.2f\n", calcularTotal()));
        factura.append("════════════════════════════════════════════════\n");
        return factura.toString();
    }

    public String toRegistro() {
        return numeroFactura + "|" + producto + "|" + cantidad + "|" + precio + "|" + fecha;
    }
}

